File: app/src/main/java/com/daniel/mobilepauker2/model/xmlsupport/FlashCardXMLPullFeedParser.java
package com.daniel.mobilepauker2.model.xmlsupport;

import android.util.SparseLongArray;
import android.util.Xml;

import com.daniel.mobilepauker2.model.FlashCard;
import com.daniel.mobilepauker2.model.pauker_native.Batch;
import com.daniel.mobilepauker2.model.pauker_native.Card;
import com.daniel.mobilepauker2.model.pauker_native.ComponentOrientation;
import com.daniel.mobilepauker2.model.pauker_native.Font;
import com.daniel.mobilepauker2.model.pauker_native.Lesson;
import com.daniel.mobilepauker2.model.pauker_native.LongTermBatch;
import com.daniel.mobilepauker2.utils.Log;

import org.xmlpull.v1.XmlPullParser;

import java.io.EOFException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FlashCardXMLPullFeedParser extends FlashCardBaseFeedParser {

    public FlashCardXMLPullFeedParser(URL feedUrl) {
        super(feedUrl);
    }

    public Lesson parse() throws EOFException {

        List<FlashCard> flashCards = null;
        XmlPullParser parser = Xml.newPullParser();
        int batchCount = 0;
        String description = "No Description";

        try {
            // auto-detect the encoding from the stream
            parser.setInput(this.getInputStream(), null);
            int eventType = parser.getEventType();
            FlashCard currentFlashCard = null;

            boolean SIDEA = false;
            boolean SIDEB = false;

            boolean done = false;
            while (eventType != XmlPullParser.END_DOCUMENT && !done) {

                String name;
                switch (eventType) {
                    case XmlPullParser.START_DOCUMENT:
                        flashCards = new ArrayList<>();
                        break;

                    case XmlPullParser.START_TAG:
                        name = parser.getName();

                        if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.LESSON)) {
                            String lessonFormatString = parser.getAttributeValue(null, "LessonFormat");
                            if (lessonFormatString != null) {
                                float lessonFormat = Float.parseFloat(lessonFormatString);
                                Log.d("FlashCardXMLPullFeedParser::parse", "Lesson format is " + lessonFormat);
                            }
                        } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.DESCRIPTION)) {
                            description = parser.nextText();
                            if (description == null) {
                                description = "No description";
                            }
                        } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.CARD)) {
                            currentFlashCard = new FlashCard();

                        } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.BATCH)) {
                            batchCount++;

                        } else if (currentFlashCard != null) {
                            currentFlashCard.setInitialBatch(batchCount - 1); // This line is repeated many times!
                            if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.FRONTSIDE) || name.equalsIgnoreCase(FlashCardXMLPullFeedParser.REVERSESIDE)) {
                                String orientation = parser.getAttributeValue(null, "Orientation");
                                String repeatByTyping = parser.getAttributeValue(null, "RepeatByTyping");
                                String learnedTimestamp = parser.getAttributeValue(null, "LearnedTimestamp");

                                currentFlashCard.setRepeatByTyping(repeatByTyping.equals("true"));

                                if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.FRONTSIDE)) {
                                    SIDEA = true;
                                    SIDEB = false;

                                    if (orientation != null) {
                                        currentFlashCard.getFrontSide().setOrientation(new ComponentOrientation("LTR"));
                                    }

                                    if (learnedTimestamp != null) {
                                        long l = Long.parseLong(learnedTimestamp.trim());
                                        currentFlashCard.getFrontSide().setLearnedTimestamp(l);
                                    }

                                } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.REVERSESIDE)) {
                                    SIDEA = false;
                                    SIDEB = true;

                                    if (orientation != null) {
                                        currentFlashCard.getReverseSide().setOrientation(new ComponentOrientation("LTR"));
                                    }

                                    if (learnedTimestamp != null) {
                                        long l = Long.parseLong(learnedTimestamp.trim());
                                        currentFlashCard.getReverseSide().setLearnedTimestamp(l);
                                    }
                                }

                                //Log.d("FlashCardXMLPullFeedParser::parse", "orientation=" + orientation);

                            } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.TEXT)) {
                                if (SIDEA) {
                                    currentFlashCard.setSideAText(parser.nextText());
                                    //Log.d("FlashCardXMLPullFeedParser::parse","sideA=" + currentFlashCard.getSideAText());

                                } else if (SIDEB) {
                                    currentFlashCard.setSideBText(parser.nextText());

                                } else {
                                    currentFlashCard.setSideAText("Empty");
                                    currentFlashCard.setSideBText("Empty");
                                }


                                //Log.d("FlashCardXMLPullFeedParser::parse","sideB=" + currentFlashCard.getSideBText());
                            } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.FONT)) {
                                String background = parser.getAttributeValue(null, "Background");
                                String bold = parser.getAttributeValue(null, "Bold");
                                String family = parser.getAttributeValue(null, "Family");
                                String foreground = parser.getAttributeValue(null, "Foreground");
                                String italic = parser.getAttributeValue(null, "Italic");
                                String size = parser.getAttributeValue(null, "Size");

                                // Set to defaults if null
                                if (background == null) {
                                    background = "-1";
                                }

                                if (bold == null) {
                                    bold = "false";
                                }

                                if (family == null) {
                                    family = "Dialog";
                                }

                                if (foreground == null) {
                                    foreground = "-16777216";
                                }

                                if (italic == null) {
                                    italic = "false";
                                }

                                if (size == null) {
                                    size = "12";
                                }

                                if (SIDEA) {
                                    currentFlashCard.getFrontSide().setFont(new Font(background,
                                            bold,
                                            family,
                                            foreground,
                                            italic,
                                            size));
                                } else if (SIDEB) {
                                    currentFlashCard.getReverseSide().setFont(new Font(background,
                                            bold,
                                            family,
                                            foreground,
                                            italic,
                                            size));
                                }
                            }
                        }
                        break;

                    case XmlPullParser.END_TAG:
                        name = parser.getName();
                        if (name.equalsIgnoreCase(CARD) && currentFlashCard != null) {
                            flashCards.add(currentFlashCard);
                        } else if (name.equalsIgnoreCase(LESSON)) {
                            done = true;
                        }
                        break;
                }
                eventType = parser.next();
            }

            return setupLesson(flashCards, description);

        } catch (Exception e) {
            Log.e("FlashCardXMLPullFeedParser:parse()", e.getMessage(), e);
            throw new RuntimeException(e);
        }
    }

    private Lesson setupLesson(List<FlashCard> flashCardList, String description) {
        Lesson newLesson = new Lesson();
        Batch summaryBatch = newLesson.getSummaryBatch();

        newLesson.setDescription(description);

        for (int i = 0; i < flashCardList.size(); i++) {
            FlashCard flashCard = flashCardList.get(i);

            if (flashCard.getInitialBatch() < 3) {
                flashCard.setLearned(false);
            } else {
                flashCard.getFrontSide().setLearned(true); // Warning using flash card set learned here sets the learned timestamp!
            }

            if (newLesson.getNumberOfLongTermBatches() < (flashCard.getInitialBatch() - 2)) {
                Log.d("FC~XMLPullFeedParser::setupLesson", "num of long term batches=" + newLesson.getNumberOfLongTermBatches());
                Log.d("FC~XMLPullFeedParser::setupLesson", "card initla batch=" + flashCard.getInitialBatch());

                int batchesToAdd = (flashCard.getInitialBatch() - 2) - newLesson.getNumberOfLongTermBatches();
                Log.d("FC~XMLPullFeedParser::setupLesson", "batchsToAdd" + batchesToAdd);

                for (int j = 0; j < batchesToAdd; j++) {
                    newLesson.addLongTermBatch();
                }
            }

            Batch batch;
            if (flashCard.isLearned()) {
                // must put the card into the corresponding long
                // term batch
                batch = newLesson.getLongTermBatch(
                        flashCard.getInitialBatch() - 3);
            } else {
                // must put the card into the unlearned batch
                batch = newLesson.getUnlearnedBatch();
            }
            batch.addCard(flashCard);
            summaryBatch.addCard(flashCard);

            //Log.d("FlashCardXMLPullFeedParser::setupLesson" , "Added card to batch " + flashCard.getInitialBatch());
        }

        //newLesson.trim();
        newLesson.refreshExpiration();
        printLessonToDebug(newLesson);

        return newLesson;
    }

    /**
     * Findet das nächste Ablaufdatum. Falls keines gefunden wird, wird {@link Long#MIN_VALUE}
     * zurückgegeben.
     * @return Eine Map mit dem frühesten Ablaufdatum <b>(index = 0)</b> und die Anzahl abgelaufener
     * Karten (<b>index = 2)</b>
     * @throws EOFException .
     */
    public SparseLongArray getNextExpireDate() throws EOFException {

        final XmlPullParser parser = Xml.newPullParser();
        final SparseLongArray map = new SparseLongArray(2);
        final long currentTimestamp = System.currentTimeMillis();

        try {
            // auto-detect the encoding from the stream
            parser.setInput(this.getInputStream(), null);
            int eventType = parser.getEventType();
            long nextExpireTimeStamp = Long.MIN_VALUE;
            int batchCount = 0;
            long expiredCards = 0;

            boolean done = false;
            while (eventType != XmlPullParser.END_DOCUMENT && !done) {

                String name;
                switch (eventType) {
                    case XmlPullParser.START_TAG:
                        name = parser.getName();

                        if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.LESSON)) {
                            String lessonFormatString = parser.getAttributeValue(null, "LessonFormat");
                            if (lessonFormatString != null) {
                                float lessonFormat = Float.parseFloat(lessonFormatString);
                                Log.d("FlashCardXMLPullFeedParser::parse", "Lesson format is " + lessonFormat);
                            }
                        } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.BATCH)) {
                            batchCount++;
                        } else if (name.equalsIgnoreCase(FlashCardXMLPullFeedParser.FRONTSIDE)
                                || name.equalsIgnoreCase(FlashCardXMLPullFeedParser.REVERSESIDE)) {
                            String learnedTimestamp = parser.getAttributeValue(null, "LearnedTimestamp");

                            if (learnedTimestamp != null) {
                                double factor = Math.pow(Math.E, batchCount - 4);
                                long expirationTime = (long) (LongTermBatch.getExpirationUnit() * factor);
                                try {
                                    long expireTimeStamp = Long.parseLong(learnedTimestamp) + expirationTime;
                                    if (nextExpireTimeStamp == Long.MIN_VALUE
                                            || expireTimeStamp < nextExpireTimeStamp) {
                                        nextExpireTimeStamp = expireTimeStamp;
                                    }

                                    long diff = currentTimestamp - Long.parseLong(learnedTimestamp);
                                    if (diff > expirationTime) {
                                        expiredCards++;
                                    }

                                } catch (NumberFormatException ignored) {}
                            }

                        }
                        break;

                    case XmlPullParser.END_TAG:
                        name = parser.getName();
                        if (name.equalsIgnoreCase(LESSON)) {
                            done = true;
                        }
                        break;
                }
                eventType = parser.next();
            }

            map.put(0, nextExpireTimeStamp);
            map.put(1, expiredCards);
            return map;

        } catch (Exception e) {
            Log.e("FlashCardXMLPullFeedParser:parse()", e.getMessage(), e);
            throw new RuntimeException(e);
        }
    }

    private void printLessonToDebug(Lesson lesson) {
        Collection<Card> cards = lesson.getLearnedCards();
        Log.d("FlashCardXMLPullFeedParser::setupLesson", "Size of learned cards is " + cards.size());

        cards = lesson.getExpiredCards();
        Log.d("FlashCardXMLPullFeedParser::setupLesson", "Size of expired cards is " + cards.size());

        cards = lesson.getShortTermList();
        Log.d("FlashCardXMLPullFeedParser::setupLesson", "Size of shortTerm cards is " + cards.size());

        cards = lesson.getCards();
        Log.d("FlashCardXMLPullFeedParser::setupLesson", "Size of all cards is " + cards.size());

        cards = lesson.getUltraShortTermList();
        Log.d("FlashCardXMLPullFeedParser::setupLesson", "Size of ultraShortTerm cards is " + cards.size());

        Log.d("FlashCardXMLPullFeedParser::setupLesson", "Number of longterm batches is" + lesson.getLongTermBatches().size());

    }
}

File: app/src/main/java/com/daniel/mobilepauker2/core/model/xmlsupport/FlashCardXMLPullFeedParser.kt
package com.daniel.mobilepauker2.core.model.xmlsupport

import android.util.SparseLongArray
import android.util.Xml
import com.daniel.mobilepauker2.pauker_native.FlashCard
import com.daniel.mobilepauker2.pauker_native.*
import com.daniel.mobilepauker2.core.Constants
import com.daniel.mobilepauker2.pauker_native.Log
import org.xmlpull.v1.XmlPullParser
import java.net.URL
import java.util.*
import kotlin.math.pow

class FlashCardXMLPullFeedParser(feedUrl: URL) : FlashCardBaseFeedParser(feedUrl) {

    @Suppress("LocalVariableName")
    override fun parse(): Lesson {
        var flashCards: MutableList<FlashCard>? = null
        val parser = Xml.newPullParser()
        var batchCount = 0
        var description: String? = "No Description"
        return try { // auto-detect the encoding from the stream
            parser.setInput(this.inputStream, null)
            var eventType = parser.eventType
            var currentFlashCard: FlashCard? = null
            var SIDEA = false
            var SIDEB = false
            var done = false
            while (eventType != XmlPullParser.END_DOCUMENT && !done) {
                var name: String
                when (eventType) {
                    XmlPullParser.START_DOCUMENT -> flashCards = ArrayList()
                    XmlPullParser.START_TAG -> {
                        name = parser.name
                        if (name.equals(LESSON, ignoreCase = true)) {
                            val lessonFormatString =
                                parser.getAttributeValue(null, "LessonFormat")
                            if (lessonFormatString != null) {
                                val lessonFormat = lessonFormatString.toFloat()
                                Log.d(
                                    "FlashCardXMLPullFeedParser::parse",
                                    "Lesson format is $lessonFormat"
                                )
                            }
                        } else if (name.equals(DESCRIPTION, ignoreCase = true)) {
                            description = parser.nextText()
                            if (description == null) {
                                description = "No description"
                            }
                        } else if (name.equals(CARD, ignoreCase = true)) {
                            currentFlashCard =
                                FlashCard()
                        } else if (name.equals(BATCH, ignoreCase = true)) {
                            batchCount++
                        } else if (currentFlashCard != null) {
                            currentFlashCard.initialBatch =
                                batchCount - 1 // This line is repeated many times!
                            if (name.equals(FRONTSIDE, ignoreCase = true)
                                || name.equals(REVERSESIDE, ignoreCase = true)
                            ) {
                                var orientation =
                                    parser.getAttributeValue(null, "Orientation")
                                orientation = orientation ?: Constants.STANDARD_ORIENTATION
                                val repeatByTyping =
                                    parser.getAttributeValue(null, "RepeatByTyping")
                                val bRepeatByTyping =
                                    if (repeatByTyping == null) Constants.STANDARD_REPEAT else repeatByTyping == "true"
                                val learnedTimestamp =
                                    parser.getAttributeValue(null, "LearnedTimestamp")
                                if (name.equals(FRONTSIDE, ignoreCase = true)) {
                                    SIDEA = true
                                    SIDEB = false
                                    currentFlashCard.frontSide.orientation =
                                        ComponentOrientation(
                                            orientation
                                        )
                                    currentFlashCard.setRepeatByTyping(bRepeatByTyping)
                                    if (learnedTimestamp != null) {
                                        val l =
                                            learnedTimestamp.trim { it <= ' ' }.toLong()
                                        currentFlashCard.setLearnedTimeStamp(l)
                                    }
                                } else if (name.equals(REVERSESIDE, ignoreCase = true)) {
                                    SIDEA = false
                                    SIDEB = true
                                    currentFlashCard.reverseSide.orientation =
                                        ComponentOrientation(
                                            orientation
                                        )
                                    currentFlashCard.reverseSide
                                        .setRepeatByTyping(bRepeatByTyping)
                                    if (learnedTimestamp != null) {
                                        val l =
                                            learnedTimestamp.trim { it <= ' ' }.toLong()
                                        currentFlashCard.reverseSide.setLearnedTimeStamp(l)
                                    }
                                }
                                //Log.d("FlashCardXMLPullFeedParser::parse", "orientation=" + orientation);
                            } else if (name.equals(TEXT, ignoreCase = true)) {
                                when {
                                    SIDEA -> {
                                        currentFlashCard.sideAText = parser.nextText()
                                        //Log.d("FlashCardXMLPullFeedParser::parse","sideA=" + currentFlashCard.getSideAText());
                                    }
                                    SIDEB -> {
                                        currentFlashCard.sideBText = parser.nextText()
                                    }
                                    else -> {
                                        currentFlashCard.sideAText = "Empty"
                                        currentFlashCard.sideBText = "Empty"
                                    }
                                }
                                //Log.d("FlashCardXMLPullFeedParser::parse","sideB=" + currentFlashCard.getSideBText());
                            } else if (name.equals(FONT, ignoreCase = true)) {
                                var background =
                                    parser.getAttributeValue(null, "Background")
                                var bold = parser.getAttributeValue(null, "Bold")
                                var family =
                                    parser.getAttributeValue(null, "Family")
                                var foreground =
                                    parser.getAttributeValue(null, "Foreground")
                                var italic =
                                    parser.getAttributeValue(null, "Italic")
                                var size = parser.getAttributeValue(null, "Size")
                                // Set to defaults if null
                                if (background == null) {
                                    background = "-1"
                                }
                                if (bold == null) {
                                    bold = "false"
                                }
                                if (family == null) {
                                    family = "Dialog"
                                }
                                if (foreground == null) {
                                    foreground = "-16777216"
                                }
                                if (italic == null) {
                                    italic = "false"
                                }
                                if (size == null) {
                                    size = "12"
                                }
                                if (SIDEA) {
                                    currentFlashCard.frontSide.font = Font(
                                        background,
                                        bold,
                                        family,
                                        foreground,
                                        italic,
                                        size
                                    )
                                } else if (SIDEB) {
                                    currentFlashCard.reverseSide.font = Font(
                                        background,
                                        bold,
                                        family,
                                        foreground,
                                        italic,
                                        size
                                    )
                                }
                            }
                        }
                    }
                    XmlPullParser.END_TAG -> {
                        name = parser.name
                        if (name.equals(
                                CARD,
                                ignoreCase = true
                            ) && currentFlashCard != null
                        ) {
                            flashCards!!.add(currentFlashCard)
                        } else if (name.equals(
                                LESSON,
                                ignoreCase = true
                            )
                        ) {
                            done = true
                        }
                    }
                }
                eventType = parser.next()
            }
            setupLesson(flashCards, description)
        } catch (e: Exception) {
            Log.e("FlashCardXMLPullFeedParser:parse()", e.message, e)
            throw RuntimeException(e)
        }
    }

    private fun setupLesson(flashCardList: List<FlashCard>?, description: String?): Lesson {
        val newLesson = Lesson()
        val summaryBatch: Batch? = newLesson.summaryBatch
        newLesson.description = description

        for (i in flashCardList!!.indices) {
            val flashCard = flashCardList[i]
            if (flashCard.initialBatch < 3) {
                flashCard.isLearned = false
            } else {
                flashCard.frontSide.isLearned =
                    true // Warning using flash card set learned here sets the learned timestamp!
            }

            if (newLesson.numberOfLongTermBatches < flashCard.initialBatch - 2) {
                Log.d(
                    "FC~XMLPullFeedParser::setupLesson",
                    "num of long term batches=" + newLesson.numberOfLongTermBatches
                )
                Log.d(
                    "FC~XMLPullFeedParser::setupLesson",
                    "card initla batch=" + flashCard.initialBatch
                )
                val batchesToAdd =
                    flashCard.initialBatch - 2 - newLesson.numberOfLongTermBatches
                Log.d(
                    "FC~XMLPullFeedParser::setupLesson",
                    "batchsToAdd$batchesToAdd"
                )
                for (j in 0 until batchesToAdd) {
                    newLesson.addLongTermBatch()
                }
            }

            val batch =
                if (flashCard.isLearned) {
                    // must put the card into the corresponding long term batch
                    newLesson.getLongTermBatch(flashCard.initialBatch - 3)
                } else { // must put the card into the unlearned batch
                    newLesson.unlearnedBatch
                }

            batch.addCard(flashCard)
            summaryBatch!!.addCard(flashCard)
        }
        newLesson.refreshExpiration()
        printLessonToDebug(newLesson)
        return newLesson
    }// auto-detect the encoding from the stream

    /**
     * Findet das nächste Ablaufdatum. Falls keines gefunden wird, wird [Long.MIN_VALUE]
     * zurückgegeben.
     * @return Eine Map mit dem frühesten Ablaufdatum **(index = 0)** und die Anzahl abgelaufener
     * Karten (**index = 1)**
     */
    val nextExpireDate: SparseLongArray
        get() {
            val parser = Xml.newPullParser()
            val map = SparseLongArray(2)
            val currentTimestamp = System.currentTimeMillis()
            return try { // auto-detect the encoding from the stream
                parser.setInput(this.inputStream, null)
                var eventType = parser.eventType
                var nextExpireTimeStamp = Long.MIN_VALUE
                var batchCount = 0
                var expiredCards: Long = 0
                var done = false
                while (eventType != XmlPullParser.END_DOCUMENT && !done) {
                    var name: String
                    when (eventType) {
                        XmlPullParser.START_TAG -> {
                            name = parser.name
                            if (name.equals(LESSON, ignoreCase = true)) {
                                val lessonFormatString =
                                    parser.getAttributeValue(null, "LessonFormat")
                                if (lessonFormatString != null) {
                                    val lessonFormat = lessonFormatString.toFloat()
                                    Log.d(
                                        "FlashCardXMLPullFeedParser::parse",
                                        "Lesson format is $lessonFormat"
                                    )
                                }
                            } else if (name.equals(BATCH, ignoreCase = true)) {
                                batchCount++
                            } else if (name.equals(FRONTSIDE, ignoreCase = true)
                                || name.equals(REVERSESIDE, ignoreCase = true)
                            ) {
                                val learnedTimestamp =
                                    parser.getAttributeValue(null, "LearnedTimestamp")
                                if (learnedTimestamp != null) {
                                    val factor = Math.E.pow(batchCount - 4.toDouble())
                                    val expirationTime =
                                        (LongTermBatch.expirationUnit * factor).toLong()
                                    try {
                                        val expireTimeStamp =
                                            learnedTimestamp.toLong() + expirationTime
                                        if (nextExpireTimeStamp == Long.MIN_VALUE
                                            || expireTimeStamp < nextExpireTimeStamp
                                        ) {
                                            nextExpireTimeStamp = expireTimeStamp
                                        }
                                        val diff =
                                            currentTimestamp - learnedTimestamp.toLong()
                                        if (diff > expirationTime) {
                                            expiredCards++
                                        }
                                    } catch (ignored: NumberFormatException) {
                                    }
                                }
                            }
                        }
                        XmlPullParser.END_TAG -> {
                            name = parser.name
                            if (name.equals(LESSON, ignoreCase = true)) {
                                done = true
                            }
                        }
                    }
                    eventType = parser.next()
                }
                map.put(0, nextExpireTimeStamp)
                map.put(1, expiredCards)
                map
            } catch (e: Exception) {
                Log.e("FlashCardXMLPullFeedParser:parse()", e.message, e)
                throw RuntimeException(e)
            }
        }

    private fun printLessonToDebug(lesson: Lesson) {
        var cards = lesson.learnedCards
        Log.d(
            "FlashCardXMLPullFeedParser::setupLesson",
            "Size of learned cards is " + cards.size
        )
        cards = lesson.expiredCards
        Log.d(
            "FlashCardXMLPullFeedParser::setupLesson",
            "Size of expired cards is " + cards.size
        )
        cards = lesson.shortTermList
        Log.d(
            "FlashCardXMLPullFeedParser::setupLesson",
            "Size of shortTerm cards is " + cards.size
        )
        cards = lesson.cards
        Log.d(
            "FlashCardXMLPullFeedParser::setupLesson",
            "Size of all cards is " + cards.size
        )
        cards = lesson.ultraShortTermList
        Log.d(
            "FlashCardXMLPullFeedParser::setupLesson",
            "Size of ultraShortTerm cards is " + cards.size
        )
        Log.d(
            "FlashCardXMLPullFeedParser::setupLesson",
            "Number of longterm batches is" + lesson.longTermBatches.size
        )
    }
}